Allow non-exclusive access to nodes

WilliamMcCumstie opened the pull request:
This PR refactors the `AllocationRegistry` to handle non-exclusive access to nodes. There are current three types of "shared resources" that are used to manage the non-exclusive access: `cpus`, `gpus`, and `memory`. The maximum available "shared resources" was implemented with the "profiler" work and is stored on each `node` (technically a `NodeAttribute` object).

For the first pass for "non-exclusive" access, it is assumed that the allocation will be "homogenous" across all nodes. This means the allocated shared resources per node are the same for each node within the allocation. This allows the `cpus_per_node`, `gpus_per_node`, and `memory_per_node` fields to be stored on the `Job` object. Allowing for "heterogeneous" allocations was considered out of scope for the initial implementation.

The `AllocationRegistry` required a fair amount of refactoring as it previously assumed that allocations-nodes existed in a 0..1-1 relationship. Non exclusive access breaks this assumption as resources on a single node could be shared across many jobs vis-a-vis with allocations. For this reason the `AllocationRegistry#for_node` method has been updated to support multiple per node.

It is important that the allocations are still tracked on a per node basis, as this is the underlining assumption of the `EventProcessor` and `Submission` classes. To changing the `AllocationRegistry` to primarily track non-node resources, would have implications for how jobs are submitted, cancelled, and deallocated.

Exclusive access is managed by a simple `exclusive` flag on the `Job`. This proved to be a fair easier implementation then trying to consumer all the resources on the assigned nodes.

Non-exclusive access has also identified a defect in how jobs are load balanced. Previously the next available node would be selected, which works fine for exclusive access. However non-exclusive access means the first node would be continually selected until most/ all of its resources have been allocated. Load balancing is a problem which will need to be addressed by itself, however a basic `LoadBalancer` has been added.

Finally non-exclusive access is linked to the idea of `NTASKS` which is currently always one. However once `NTASKS` is implemented, it is feasible that a "two task job" maybe assigned to a single node twice.

> Example: If a two task job requires two cores per task, and there are many quad core machines, than the job maybe assigned all the cores on a single machine instead of two cores on two machines.

In preparation, the `AllocationRegistry` implements a `max_parallel_per_node` method which determines how many times a job could be ran in parrallel on the node before it consumes all the available resources.

WilliamMcCumstie committed:
Refactor AllocationRegistry to allow non-exclusivity

Previously the allocation registry only allowed a node to
have a single allocation

With the change to non-exclusivity (TBA) this will no longer
be the case

The registry has been refactored to return the allocation for
a node as an array

WilliamMcCumstie committed:
Switch AllocationRegistry to use a ReadWriteLock

WilliamMcCumstie committed:
Remove rogue comment about NTASKS

That is not how NTASKS work

WilliamMcCumstie committed:
Start implementing *_per_node attributes

WilliamMcCumstie committed:
Factor in pre-existing allocation for non-exclusivity

WilliamMcCumstie committed:
Consider memory and gpus as part of the allocation

WilliamMcCumstie committed:
Account for exclusive jobs when allocating

WilliamMcCumstie committed:
Update AllocationRegistry spec to test #add

WilliamMcCumstie committed:
Ensure the allocation registry does not oversubscribe

WilliamMcCumstie committed:
Add a basic load balancer

WilliamMcCumstie committed:
Remove the old partition methods for selecting nodes

WilliamMcCumstie committed:
Allow the client to set the *_per_node job fields

WilliamMcCumstie committed:
Fix bug in determining a node's state

WilliamMcCumstie committed:
Remove the unnecessary clearing of allocation registry

A new `subject` is created on each test run, so there is no
need to clear it

This looks like a holdover from a previous implementation

WilliamMcCumstie committed:
Start replacing AllocationRegistry#delete

Because the AllocationRegistry maintains the node to allocation
mapping entirely, the deallocation must also be done internally

WilliamMcCumstie committed:
Allow node allocations to be remove individually

WilliamMcCumstie committed:
Handle deallocating nodes from missing jobs

WilliamMcCumstie committed:
Remove the AllocationRegistry#delete method

It is not used in the code base anymore* and has been replaced
with deallocate_node_from_job

It was used in the FiFo scheduler spec, however this has been
replaced

benarmston commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -36,86 +38,146 @@
-# 3. Adding +Allocation+s to the +AllocationRegistry+ is atomic; and therefore
-#    thread safe.
+# 3. Adding +Allocation+s to the +AllocationRegistry+ is done in write lock;
+#    and therefore thread safe.
```
Removing the documented guarantee that the `add` method is atomic worries me.  Having that guarantee prevents nodes erroneously being allocated more times than they actually support.  The method is still atomic and that guarantee is what should be documented rather than the particular mechanism used to achieve it.

benarmston commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -36,86 +38,146 @@
+    KEY_MAP.reduce(nil) do |max, (node_key, job_key)|
+      dividor = job.send(job_key).to_i
+      next max if dividor < 1
+      current = node.send(node_key).to_i / dividor - allocated[job_key]
```
This is incorrect.  Consider a node with 4 CPUs currently running a job requiring 2 CPUs and we want to know it can run another node requiring 2 CPUs.  We end up with  `current = 4 / 2 - 2` which sets `current == 0`.  Obviously that is not correct.  Instead we should have `current = ( node.send(node_key).to_i - allocated[job_key] ) / dividor`.  Which gives `(4 - 2) / 2` setting `current == 2`.

WilliamMcCumstie replied:
Maths is hard

benarmston commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -36,86 +38,146 @@
+      current = node.send(node_key).to_i / dividor - allocated[job_key]
+
+      break 0 if current < 1
+      break 1 if current == 1
```
I don't think this is right either.  Consider a quiet node with 4 CPUs and 1 GPU.  We want to know if it can run a job requiring 4 CPUs and 2 GPUs.  The answer ought to be no.  However, the CPU calculation sets `current == 1` which causes this line to break out incorrectly.

benarmston committed:
Fix calculation of how many jobs can run on a node

benarmston committed:
Add some debugging

benarmston committed:
Array tasks and node allocations work correctly

benarmston committed:
Cancelling job runs scheduler allocation

Cancelling a pending job should run the allocation routine.  It could be
blocking other jobs.

benarmston committed:
Revert change to documentation

Having the guarantee that the `add` method is atomic prevents nodes
erroneously being allocated more times than they actually support. The
method is still atomic and that guarantee is what should be documented
rather than the particular mechanism used to achieve it.

benarmston reviewed (commented):
I've added commits to this PR which address the issues I've commented on.  I decided that it was easier just to fix the issues and test that there weren't any others being uncovered.

> Previously the next available node would be selected, which works fine for exclusive access. However non-exclusive access means the first node would be continually selected until most/ all of its resources have been allocated.

Its entirely possible that the previous behaviour is the desired behaviour.  At least in some situations.

Consider when we add support for dynamically adding cloud nodes to a cluster when it is under heavy load.  We will also want to remove cloud nodes from the cluster when the load reduces.  This results in a cost/performance tradeoff.  The previous algorithm would have unconditionally preferred cost over performance.  The algorithm in this PR unconditionally prefers performance over cost.  Obviously neither would be unconditionally correct given dynamic resizing of the cluster.

Having a `LoadBalancer` class to handle this seems like a good idea.

benarmston commented:
LGTM.
